core/chains/evm/txm/txm.go

```diff
@@ -0,0 +1,462 @@
+	for {
+		_, unconfirmedCount, err := t.txStore.FetchUnconfirmedTransactionAtNonceWithCount(ctx, 0, address)
+		if err != nil {
+			return false, err
+		}
+
+		// Optimistically send up to maxInFlightSubset of the maxInFlightTransactions. After that threshold, broadcast more cautiously
+		// by checking the pending nonce so no more than maxInFlightSubset can get stuck simultaneously i.e. due
+		// to insufficient balance. We're making this trade-off to avoid storing stuck transactions and making unnecessary
+		// RPC calls. The upper limit is always maxInFlightTransactions regardless of the pending nonce.
+		if unconfirmedCount >= maxInFlightSubset {
+			if unconfirmedCount > maxInFlightTransactions {
+				t.lggr.Warnf("Reached transaction limit: %d for unconfirmed transactions", maxInFlightTransactions)
+				return true, nil
+			}
+			pendingNonce, e := t.client.PendingNonceAt(ctx, address)
+			if e != nil {
+				return false, e
+			}
+			nonce := t.getNonce(address)
+			if nonce > pendingNonce {
+				t.lggr.Warnf("Reached transaction limit. LocalNonce: %d, PendingNonce %d, unconfirmedCount: %d",
+					nonce, pendingNonce, unconfirmedCount)
+				return true, nil
+			}
+		}
+
+		nonce := t.getNonce(address)
+		tx, err := t.txStore.UpdateUnstartedTransactionWithNonce(ctx, address, nonce)
+		if err != nil {
+			return false, err
+		}
+		if tx == nil {
+			return false, nil
+		}
+		t.setNonce(address, nonce+1)
+
+		if err := t.createAndSendAttempt(ctx, tx, address); err != nil {
+			return true, err
+		}
+	}
+}
+
+func (t *Txm) createAndSendAttempt(ctx context.Context, tx *types.Transaction, address common.Address) error {
+	attempt, err := t.attemptBuilder.NewAttempt(ctx, t.lggr, tx, t.config.EIP1559)
+	if err != nil {
+		return err
+	}
+
+	if tx.Nonce == nil {
+		return fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+	}
+	if err = t.txStore.AppendAttemptToTransaction(ctx, *tx.Nonce, address, attempt); err != nil {
+		return err
+	}
+
+	return t.sendTransactionWithError(ctx, tx, attempt, address)
+}
+
+func (t *Txm) sendTransactionWithError(ctx context.Context, tx *types.Transaction, attempt *types.Attempt, address common.Address) (err error) {
+	if tx.Nonce == nil {
+		return fmt.Errorf("nonce for txID: %v is empty", tx.ID)
+	}
+	start := time.Now()
+	txErr := t.client.SendTransaction(ctx, tx, attempt)
+	tx.AttemptCount++
+	t.lggr.Infow("Broadcasted attempt", "tx", tx, "attempt", attempt, "duration", time.Since(start), "txErr: ", txErr)
+	if txErr != nil && t.errorHandler != nil {
+		if err = t.errorHandler.HandleError(tx, txErr, t.attemptBuilder, t.client, t.txStore, t.setNonce, false); err != nil {
+			return
+		}
+	} else if txErr != nil {
+		pendingNonce, err := t.client.PendingNonceAt(ctx, address)
+		if err != nil {
+			return err
+		}
+		if pendingNonce <= *tx.Nonce {
+			t.lggr.Debugf("Pending nonce for txID: %v didn't increase. PendingNonce: %d, TxNonce: %d", tx.ID, pendingNonce, *tx.Nonce)
+			return nil
```

Comment:
Why return nil?
This means there was a failure, which needs to be somehow addressed right?

Reply:
Any issues will be addressed by the backfill loop. Here we return because we don't want to update the broadcast times and the issue has already been printed above. If we want custom handling for this we need to implement the `HandleError` above.